### PR TITLE
NVDAObjects.UIA: announce drag and drop target effects

### LINKED ISSUE #12271: UIA handler: introduce UIA drag/drop event handlers for use with announcing items that were rearranged on screen
Hi,

I believe the following was asked before, but figured it would be best to put it in a systematic way:

### Background:
When rearranging UIA objects such as Start menu tiles and Action Center items (Windows 10), UIA can raise drag drop events. This is checked by Narrator (at least) to announce the new object location. For example, when rearranging Action Center items by pressing Alt+Shift+left/right arrow keys, screen readers may announce the new location of the just dragged object. This feature will become useful in future Windows 10 releases as latest Windows Insider Preview build allows one to rearragne virtual desktops and UIA drag drop events are fired.

### Is your feature request related to a problem? Please describe.
Currently NVDA does not announce items or new locations when items are rearranged. Technially, it does through windows 10 app Essentials add-on, but it would be better to bring this to core.

### Describe the solution you'd like
Let NVDA listen to UIA drag/drop events and announce the rearrange item and new location by calling obj.reportFocus method.

### Describe alternatives you've considered
Leave it up to add-ons to do this.

### Additional context
For a practical example and justifications, see:
https://blogs.windows.com/windows-insider/2021/03/17/announcing-windows-10-insider-preview-build-21337/

### Implementation strategy:
UIA drag/drop event handler code from Windows 10 App Essentials add-on will be ported to NVDA Core and will correspond to gain focus event unless folks believe a separate set of UIA events should be created. There are six events related to drag and drop, but the events we are interested in are "drag dropped" and "drag complete".

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
Hi,

This is more of a second attempt at #12271 

### Link to issue number:
Fixes #12271 
Follow-up and a different approach to #12428 

### Summary of the issue:
When drag and drop target events happen, NVDA does not announce their effects such as screen content changes such as effect of rearranging Start menu tiles.

### Description of user facing changes
NVDA will announce effects of drag and drop events in places such as:

* Windows 10 Start menu: after moving tiles
* Windows 10 Action center/Windows 11 quick setings: afterreordering settings items
* Windows 11 task view: rearranging virtual desktops
* Windows 11 Start menu: rearranging pinned items, and from Version 22H2 onwards, creating app folders

### Description of development approach
While this PR may look similar to #12428, the internals are different:

* In UIA handler, added property events for drag drop and drop target effects.
* In UIA objects, add property events for drag drop and drop target effects and announce effect text.
* For drop target effect, because drop target event may come from somewhere other than the object that "raises" it, traverse focus ancestors looking for UIA elements with drop target effect text set.
* Back in UIA handler, because drop target effect event can come from somewhere other than the focused element or its parent, treat it as a global event.

### Testing strategy:
Manual and add-on based testing:

Testing drag drop effect (prerequisite: Windows 10):

1. After building this PR from source, open Windows 10 Start menu.
2. Press Shift+Tab to go to tiles.
3. Press Alt+Shift+arrow keys to arrange tiles.
Before the PR: NVDA says nothing about drag drop effect.
After the PR: NVDA will announce the new position of the dragged tile.

Testing drop target effect (for best results, use Windows 11):

1. After building the PR from source, open multiple virtual desktops (Control+Windows+D).
2. Move to a virtual desktop (say, Desktop 1).
3. Open task view (Windows+Ta), then move to desktops list.
4. Press Alt+Shift+left/right arrows to reoder virtual desktops.

Before the PR: NVDA does not say anything about virtual desktop position.
After the PR: NVDA announces effect of reordering virtual desktop such as new item position.

### Known issues with pull request:
This PR takes care of drag and drop target effect texts announcement, but does not include additional work for handling drag start/cancel/complete and drop target enter/leave/dropped events themselves (at least drag start event and the resulting dragging state change is handled by #14097). Should the need arise to handle actual drag and drop target events set, foundation is there to support them.

### Change log entries:
Similar to #14097 but could be expanded to say (new features or bug fixes):

NVDA will announce effects of dragging items on screen in places such as rearranging Windows 10 Start menu tiles and virtual desktops in Windows 11. (#12271)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by josephsl** [comment]
CC @mazen428

File: source/UIAHandler/__init__.py [hunk 1]
```diff
@@ -169,19 +169,22 @@
...
+	UIA.UIA_ControllerForPropertyId: "UIA_controllerFor",
+	UIA.UIA_ItemStatusPropertyId: "UIA_itemStatus",
+	UIA.UIA_DragDropEffectPropertyId: "UIA_dragDropEffect",
+	UIA.UIA_DropTargetDropTargetEffectPropertyId: "UIA_dropTargetEffect",
 }
 
 globalEventHandlerGroupUIAPropertyIds = {
-	UIA.UIA_RangeValueValuePropertyId
+	UIA.UIA_RangeValueValuePropertyId,
+	UIA.UIA_DropTargetDropTargetEffectPropertyId
```
Comment (seanbudd) [inline source/UIAHandler/__init__.py:187]: added comma for easier diffs

```suggestion
	UIA.UIA_DropTargetDropTargetEffectPropertyId,
```

**Review by seanbudd — COMMENTED** [review]
Could system tests be added using the manual testing example?

**Comment by josephsl** [comment]
Hi, we can try coming up with system tests – note that rearranging Start menu tiles could work best as Windows Server 2019 (the one Appveyor is running on) qualifies as a Windows 10 system i.e. Windows 10 Version 1809 (will need to invest in some time learning about Robot framework). In the meantime I’ll apply the comma change. Thanks.

**Comment by seanbudd** [comment]
@josephsl - yes I think the system test would need to only run on the known versions of Windows with the behaviour of the test.
I don't think it worth re-writing the test for different versions of Windows.
This might not be worth the long term upkeep if the behaviour varies a lot across Windows versions.

**Comment by josephsl** [comment]
Hi, right – that’s what I’m afraid of, as I remember a discussion about adding Windows specific tests a while back. Thanks.

**Comment by josephsl** [comment]
Hi,

I think system tests cannot be done easily for this PR as drag and drop target effect text begins with a specific keyword ("place"). I'm sure it might be possible to detect substrings (text in str) but may need to read Robot framework docs more thoroughly. I think we might as well come up with an issue/PR pair (if it doesn't exist yet) just to deal with foundations to support Windows release detection (perhaps using sys.getwindowsversion function) and ways to detect substrings, as it can make future iterations of PR's like this easier to test locally in a more systematic way.

For anyone wishing to tackle system tests for PR's like this, you must first detect which Windows release the test system is running (perhaps using sys.getwindowsversion().platform_version), then emulate key presses (see existing system test robot files for details) and figure out what NVDA is saying or get a speech substring somehow.

Thanks.

**Review by seanbudd — APPROVED** [review]

File: source/NVDAObjects/UIA/__init__.py [hunk 2]
```diff
@@ -2055,6 +2055,35 @@ def event_UIA_itemStatus(self):
...
+		try:
+			dropTargetEffect = self._getUIACacheablePropertyValue(
+				UIAHandler.UIA_DropTargetDropTargetEffectPropertyId
+			)
+		except COMError:
+			dropTargetEffect = ""
+		# Sometimes drop target effect text is empty as it comes from a different object.
+		if not dropTargetEffect:
+			for element in reversed(api.getFocusAncestors()):
+				try:
```
Comment (leonardder) [inline source/NVDAObjects/UIA/__init__.py:2076]: Please avoid calling _getUIACacheablePropertyValue on NON-UIA objects.

```suggestion
				if not isinstance(element, NVDAObjects.UIA.UIA):
					continue
				try:
```

File: source/NVDAObjects/UIA/__init__.py [hunk 3]
```diff
@@ -2055,6 +2055,35 @@ def event_UIA_itemStatus(self):
...
+					dropTargetEffect = element._getUIACacheablePropertyValue(
+						UIAHandler.UIA_DropTargetDropTargetEffectPropertyId
+					)
+				except (COMError, AttributeError):
```
Comment (leonardder) [inline source/NVDAObjects/UIA/__init__.py:2080]: Together with the change above, this can be
```suggestion
				except COMError:
```

**Review by LeonarddeR — COMMENTED** [review]
One small thing

**Review by seanbudd — APPROVED** [review]

**Comment by josephsl** [comment]
Hi, yes, I’m using isinstance call in the add-on version – will investigate calling isinstance on self in the next round (today). Thanks for the reminder.

**Pull request merged**